### PR TITLE
Use NSCache for regex caches instead of NSMutableDictionary

### PR DESCRIPTION
**Pull request by jrose-signal**
...so that they can respond to memory pressure. This also avoids an explicit lock and try/finally dance, because NSCache can be accessed from multiple threads. (Though maybe it's just doing a lock itself.)

### CODE REVIEW

File: libPhoneNumber/NBPhoneNumberUtil.m [hunk 1]
```diff
@@ -135,50 +132,30 @@ - (NSError *)errorWithObject:(id)obj withDomain:(NSString *)domain {
 - (NSRegularExpression *)entireRegularExpressionWithPattern:(NSString *)regexPattern
                                                     options:(NSRegularExpressionOptions)options
                                                       error:(NSError **)error {
-  [_entireStringCacheLock lock];
-
-  @try {
```
Comment (michelle-signal) [inline libPhoneNumber/NBPhoneNumberUtil.m:140]: Why was this even here in the first place?

File: libPhoneNumber/NBPhoneNumberUtil.m [hunk 2]
```diff
@@ -40,11 +40,8 @@ static BOOL isNan(NSString *sourceString) {
 
 @interface NBPhoneNumberUtil ()
 
-@property(nonatomic, strong) NSLock *entireStringCacheLock;
-@property(nonatomic, strong) NSMutableDictionary *entireStringRegexCache;
-
-@property(nonatomic, strong) NSLock *lockPatternCache;
-@property(nonatomic, strong) NSMutableDictionary *regexPatternCache;
+@property(nonatomic, strong) NSCache *entireStringRegexCache;
```
Comment (michelle-signal) [inline libPhoneNumber/NBPhoneNumberUtil.m]: Since we're updating this anyway, we might as well add lightweight generics while we're here. Thoughts?

Comment (michelle-signal) [reply libPhoneNumber/NBPhoneNumberUtil.m]: There's something that also bugs me about declaring a property just to access the underlying ivar directly. I get that we're inheriting this code from elsewhere so no need to change it

**Review by michelle-signal — APPROVED** [review]
This looks mostly fine. My one real concern is dropping the `@try` block. I'm not sure what would've ever thrown here though.

File: libPhoneNumber/NBPhoneNumberUtil.m (hunk 1 above)
Comment (jrose-signal) [reply libPhoneNumber/NBPhoneNumberUtil.m:140]: To make sure the unlock happened if an invalid regex pattern sneaks in and throws an exception. Now we're trusting that NSCache handles that correctly.

File: libPhoneNumber/NBPhoneNumberUtil.m (hunk 2 above)
Comment (jrose-signal) [reply libPhoneNumber/NBPhoneNumberUtil.m]: Yeah, I figure this should stay close to the original style. I can add the generics, though.

**Pull request merged**